File: backend/app/models/alert.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin


class AlertRule(Base, TimestampMixin):
    """告警规则"""
    __tablename__ = "alert_rules"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(100))
    # 规则类型: error_rate / high_frequency / slow_query / connection_fail
    rule_type: Mapped[str] = mapped_column(String(30))
    # 阈值配置（JSON 字符串）
    # error_rate: {"window_minutes": 5, "threshold_percent": 50}
    # high_frequency: {"window_minutes": 1, "max_calls": 100}
    # slow_query: {"threshold_ms": 5000}
    # connection_fail: {"consecutive": 3}
    threshold_config: Mapped[str] = mapped_column(Text, default="{}")
    # 作用范围: global / user / datasource
    scope: Mapped[str] = mapped_column(String(20), default="global")
    target_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    # 告警抑制：同规则 N 分钟内不重复触发
    suppress_minutes: Mapped[int] = mapped_column(Integer, default=10)


class AlertRecord(Base):
    """告警记录"""
    __tablename__ = "alert_records"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    rule_id: Mapped[int] = mapped_column(Integer, index=True)
    rule_name: Mapped[str] = mapped_column(String(100))
    rule_type: Mapped[str] = mapped_column(String(30))
    # 触发详情
    detail: Mapped[str] = mapped_column(Text, default="")
    # 处理状态: pending / acknowledged / resolved
    status: Mapped[str] = mapped_column(String(20), default="pending")
    triggered_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    resolved_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

File: frontend/src/api/alert.ts
import request from './index'

export interface AlertRule {
  id: number
  name: string
  rule_type: string
  threshold_config: Record<string, number>
  scope: string
  target_id: number | null
  is_active: boolean
  suppress_minutes: number
  created_at: string | null
}

export interface AlertRecord {
  id: number
  rule_id: number
  rule_name: string
  rule_type: string
  detail: string
  status: string
  triggered_at: string | null
  resolved_at: string | null
}

export function listAlertRules(): Promise<AlertRule[]> {
  return request.get('/alerts/rules')
}

export function createAlertRule(data: Partial<AlertRule>): Promise<AlertRule> {
  return request.post('/alerts/rules', data)
}

export function updateAlertRule(id: number, data: Partial<AlertRule>): Promise<AlertRule> {
  return request.put(`/alerts/rules/${id}`, data)
}

export function deleteAlertRule(id: number): Promise<void> {
  return request.delete(`/alerts/rules/${id}`)
}

export function listAlertRecords(status?: string, limit?: number): Promise<AlertRecord[]> {
  return request.get('/alerts/records', { params: { status, limit } })
}

export function getPendingAlertCount(): Promise<{ count: number }> {
  return request.get('/alerts/records/pending-count')
}

export function acknowledgeAlert(id: number): Promise<void> {
  return request.put(`/alerts/records/${id}/acknowledge`)
}

export function resolveAlert(id: number): Promise<void> {
  return request.put(`/alerts/records/${id}/resolve`)
}
